pkcs8/src/one_asymmetric_key.rs

```diff
@@ -0,0 +1,143 @@
+///     [[2: publicKey        [1] PublicKey OPTIONAL ]],
+///     ...
+///   }
+///
+/// Version ::= INTEGER { v1(0), v2(1) } (v1, ..., v2)
+///
+/// PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+///
+/// PrivateKey ::= OCTET STRING
+///
+/// Attributes ::= SET OF Attribute
+///
+/// PublicKey ::= BIT STRING
+/// ```
+///
+/// [RFC 5958 Section 2]: https://datatracker.ietf.org/doc/html/rfc5958#section-2
+#[derive(Clone)]
+pub struct OneAsymmetricKey<'a> {
+    /// PKCS#8 document, only [`Version::V1`] and [`Version::V2`] are valid.
+    pub version: Version,
+
+    /// X.509 [`AlgorithmIdentifier`] for the private key type
+    pub algorithm: AlgorithmIdentifier<'a>,
+
+    /// Private key data
+    pub private_key: &'a [u8],
+
+    /// Public key data, optionally available if version is v2
+    pub public_key: Option<&'a [u8]>,
+}
+
+impl<'a> TryFrom<&'a [u8]> for OneAsymmetricKey<'a> {
+    type Error = Error;
+
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        Ok(Self::from_der(bytes)?)
+    }
+}
+
+impl<'a> TryFrom<der::Any<'a>> for OneAsymmetricKey<'a> {
+    type Error = der::Error;
+
+    fn try_from(any: der::Any<'a>) -> der::Result<OneAsymmetricKey<'a>> {
+        any.sequence(|decoder| {
+            // Parse and validate `version` INTEGER.
+            let version = Version::decode(decoder)?;
+
+            let algorithm = decoder.decode()?;
+            let private_key = decoder.octet_string()?.into();
+
+            let public_key: Option<&[u8]> = match &version {
+                Version::V1 => {
+                    // run once, throw away an Attributes field (for now)
+                    // TODO: Properly process and store attributes
+                    decoder.decode::<Option<_AttributesStub>>()?;
+
+                    None
+                }
+                Version::V2 => {
+                    while let Some(_) = decoder.decode::<Option<_AttributesStub>>()? {
+                        // Throw away all Attributes (for now)
+                        // TODO: Properly process and store attributes
+                    }
+
+                    let ret = if let Some(ContextualTo1(pubkey)) =
+                        decoder.decode::<Option<ContextualTo1<BitString<'a>>>>()?
+                    {
+                        Some(pubkey.as_bytes())
+                    } else {
+                        None
+                    };
+
+                    while let Some(_) = decoder.decode::<Option<ContextualTo1<BitString<'a>>>>()? {
+                        // Throw away further public keys (for now)
+                        // FIXME: the documentation says "...,",
+                        //  meaning more fields of the same type can exist,
+                        //  considering that the rest of the documentation isn't talking about "multiple public keys",
+                        //  I assume it is okay to only get the first value, and ignore the rest.
+                    }
+
+                    ret
```

Comment:
Here's a more succinct way of implementing this:

```suggestion
                    let ret = None;

                    while let Some(pk) = decoder.context_specific_optional(1, |dec| dec.bit_string())? {
                        ret.get_or_insert(pk);
                    }

                    ret
```

Reply:
With this, i'll pick the last public key found? Do you want that?

Reply:
`get_or_insert` only performs an insert if the value is `None`, so it will get the first. However...

I just noticed above that for `Version::V2`, it should *always* try to decode at least one public key, so my code example is wrong. Something closer to what you originally had is probably more correct.